content/post/2019-12-19-mutable-api.Rmarkdown

```diff
@@ -0,0 +1,184 @@
+---
+title: Objects that feel mutable in R
+date: '2019-12-26'
+slug: mutable-api
+tags:
+  - package development
+---
+
+In R, most often, to change an object, you need to re-assign its new value to it. But sometimes, things feel different because objects are mutable or it seems they are, be it in base R code or in the code of packages. Why and how provide a mutable API/interface in R code? In this blog post, we shall explore <s>few odd examples</s> a few examples to better understand mutable APIs in R.
+
+## Preamble: what do we mean by mutable?
+
+### What is mutable
+
+As explained in the [chapter "Names and values" of the Advanced R book by Hadley Wickham](https://adv-r.hadley.nz/names-values.html), in R objects are actually not mutable, because of the [copy-on-modify behaviour](https://adv-r.hadley.nz/names-values.html#copy-on-modify). The book has very clear diagrams showing how a name is bound to a binding corresponding to an object. When you think you're modifying an object, a new object with a new value has been created and bound to the initial name. The original object is unchanged, it's not mutable.
+
+Environments are mutable, they can be [modified in place](https://adv-r.hadley.nz/names-values.html#modify-in-place). 
```

Comment:
Yeah, so some objects are mutable, environments are one, external pointers are another, and maybe there is more. So I think we could probably say something about exceptions in the previous paragraphs.